include/dr/shp/algorithms/transform.hpp

```diff
@@ -0,0 +1,109 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+#pragma once
+
+#include <dr/details/alignment.hpp>
+#include <dr/shp/init.hpp>
+#include <dr/shp/util.hpp>
+
+namespace shp {
+
+namespace __detail {
+template <class T> T *get_local_pointer(shp::device_ptr<T> r) {
+  return r.local();
+}
+template <class T> T *get_local_pointer(T *t) { return t; }
+} // namespace __detail
+
+void transform(auto &&policy, lib::distributed_range auto &&in,
```

Comment:
In an earlier discussion, @BenBrock said that some things could be implemented with `zip` and `for_each`, and we would not have to partition the work. Is that possible for `transform`?

Reply:
good point, we should reuse code if possible - I will try it tomorrow

Reply:
I had been thinking of something like

```cpp
template <typename R, typename O, typename Fn>
void transform(R&& input_range, O&& output_range, Fn fn) {
  auto zipped_range = shp::views::zip(input_range, output_range);
  shp::for_each(shp::par_unseq, [](auto&& a, auto&& b) { b = fn(a); });
}
```

For a default implementation.  It is true you might want to do something more complicated when they're not aligned, but I think using `zip` makes it a lot easier, since it picks out the overlapping segments for you.

Reply:
You could then do something like

```cpp
template <typename R, typename O, typename Fn>
void transform(R&& input_range, O&& output_range, Fn fn) {
  auto r = shp::views::zip(input_range, output_range);

  for (auto&& segs : r.zipped_segments()) {
    if (segs[0].rank() == segs[1].rank()) {
      // Both segments are on same rank, just run as normal.
      for_each(...);
    } else {
      // Create local buffer on GPU
      shp::device_vector<T, ...> buffer(segs[0].size(), ...);

      // Run locally as normal
      auto e = for_each_async(...);

      // Copy result from buffer to remote
      shp::copy_async(buffer.begin(), buffer.end(), segs[1].begin(), e);
    }
  }
}
```

Reply:
> I had been thinking of something like

Yes, that would be much easier to maintain. As you show, it has a default that will give good performance for the typical case (everything aligned) and still allows special case optimizations. There are many variants of transform (https://en.cppreference.com/w/cpp/algorithm/transform) so we need to funnel all the implementations through some common implementation anyway.

Reply:
On second thought, I actually think the first implementation ought to perform well, as the memory controller should be able to properly handle the cross-GPU copies.  (At least I would expect this to perform well with P2P on Nvidia GPUs.)